protocol/java/org/openyolo/protocol/CredentialSaveRequest.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright 2017 The OpenYOLO Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.openyolo.protocol;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+import android.support.annotation.NonNull;
+
+import org.openyolo.protocol.internal.ByteStringConverters;
+import org.openyolo.protocol.internal.ClientVersionUtil;
+import org.openyolo.protocol.internal.CollectionConverter;
+import org.openyolo.protocol.internal.CredentialConverter;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.hamcrest.core.Every.everyItem;
+import static org.hamcrest.core.IsNull.notNullValue;
+import static org.openyolo.protocol.internal.CustomMatchers.notNullOrEmptyString;
+import static org.valid4j.Assertive.require;
+
+/**
+ * A request to save a given {@link Credential credential}. To make the request the client
+ * encodes it in its associated protocol buffer byte array form in an activity intent
+ * via the {@link ProtocolConstants#EXTRA_SAVE_REQUEST} extra.
+ *
+ * @see <a href="https://spec.openyolo.org/openyolo-android-spec.html#saving-credentials">
+ *     OpenYOLO specification: Saving Credentials</a>
+ */
+public final class CredentialSaveRequest implements Parcelable {
+
+    /**
+     * Parcelable reader for {@link CredentialSaveRequest} instances.
+     * @see android.os.Parcelable
+     */
+    public static final Creator<CredentialSaveRequest> CREATOR = new CredentialSaveRequestCreator();
+
+    @NonNull
+    private final Credential mCredential;
+
+    @NonNull
+    private final Map<String, byte[]> mAdditionalProperties;

```

Comment:
preferably retain Map<String, ByteString> as then the values are properly immutable.